class HELP:
    profile = 'ID of the profile as defined in your config TOML. The profile consists of Galaxy server credentials and a Galaxy Tool-ID to be used for your BLAST call'
    query = 'Path / filename of file with nucleotide query sequence(s)'
    task = 'Task type'
    db = 'Database name'
    evalue = 'Expectation value cutoff'
    out = 'Path / filename of file to store the BLAST result'
    outfmt = 'Output format'
    html = 'Format output as HTML document'
    dust = 'Filter out low complexity regions (with DUST)'
    seg = 'Filter out low complexity regions (with SEG)'
    db_gencode = 'Genetic code to use to translate database/subjects (see user manual for details)'
    matrix = 'Scoring matrix name (normally BLOSUM62)'
    max_target_seqs = 'Maximum number of aligned sequences to keep (value of 5 or more is recommended) Default = 500'
    num_descriptions = 'Number of database sequences to show one-line descriptions for. Not applicable for outfmt > 4. Default = 500 * Incompatible with:  max_target_seqs'
    num_alignments = 'Number of database sequences to show alignments for. Default = 250 * Incompatible with:  max_target_seqs'
    threshold = 'Minimum word score such that the word is added to the BLAST lookup table'
    strand = 'Query strand(s) to search against database/subject'
    max_hsps = 'Maximum number of HSPs (alignments) to keep for any single query-subject pair'
    perc_identity = 'Percent identity cutoff'
    word_size = 'Word size for wordfinder algorithm'
    ungapped = 'Perform ungapped alignment only?'
    parse_deflines = 'Should the query and subject defline(s) be parsed?'
    qcov_hsp_perc = 'Minimum query coverage per hsp (percentage, 0 to 100)'
    window_size = 'Multiple hits window size: use 0 to specify 1-hit algorithm, leave blank for default'
    gapopen = 'Cost to open a gap'
    gapextend = 'Cost to extend a gap'
    comp_based_stats = 'Use composition-based statistics: D or d: default (equivalent to 2 ); 0 or F or f: No composition-based statistics; 1: Composition-based statistics as in NAR 29:2994-3005, 2001; 2 or T or t : Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, conditioned on sequence properties; 3: Composition-based score adjustment as in Bioinformatics 21:902-911, 2005, unconditionally'
    use_sw_tback = 'Compute locally optimal Smith-Waterman alignments?'